File: tests/test_check_repo_hygiene.py
#!/usr/bin/env python3
import importlib.util
import sys
import tempfile
from pathlib import Path
import unittest


def _load_module():
    repo_root = Path(__file__).resolve().parents[1]
    module_path = repo_root / "scripts" / "check_repo_hygiene.py"
    spec = importlib.util.spec_from_file_location("check_repo_hygiene", module_path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"Unable to load module from {module_path}")
    module = importlib.util.module_from_spec(spec)
    sys.modules[spec.name] = module
    spec.loader.exec_module(module)
    return module


repo_hygiene = _load_module()


class CheckRepoHygieneTests(unittest.TestCase):
    def test_detects_default_blocked_root_file(self):
        with tempfile.TemporaryDirectory(prefix="zkfuzz_repo_hygiene_") as tmpdir:
            root = Path(tmpdir)
            (root / "new_file.txt").write_text("", encoding="utf-8")
            matches = repo_hygiene.blocked_root_files(
                root, repo_hygiene.DEFAULT_BLOCKED_ROOT_FILES
            )
            self.assertEqual(matches, ["new_file.txt"])

    def test_parse_blocklist_file_ignores_comments(self):
        with tempfile.TemporaryDirectory(prefix="zkfuzz_repo_hygiene_") as tmpdir:
            blocklist = Path(tmpdir) / "blocklist.txt"
            blocklist.write_text(
                "# comment\n\ncustom_placeholder.txt\n  extra.log  \n",
                encoding="utf-8",
            )
            blocked = repo_hygiene.parse_blocklist_file(blocklist)
            self.assertEqual(blocked, {"custom_placeholder.txt", "extra.log"})

    def test_main_passes_when_no_matches(self):
        with tempfile.TemporaryDirectory(prefix="zkfuzz_repo_hygiene_") as tmpdir:
            root = Path(tmpdir)
            (root / "README.md").write_text("ok\n", encoding="utf-8")
            exit_code = repo_hygiene.main(["--repo-root", str(root)])
            self.assertEqual(exit_code, 0)


if __name__ == "__main__":
    unittest.main()

File: tests/test_check_repo_hygiene.rs
use std::path::Path;
use tempfile::tempdir;
use zk_fuzzer::checks::repo_hygiene::{
    blocked_root_files, build_report, parse_blocklist_file, DEFAULT_BLOCKED_ROOT_FILES,
};

#[test]
fn detects_default_blocked_root_files() {
    let tmp = tempdir().expect("tempdir");
    let root = tmp.path();
    std::fs::write(root.join(".env"), "SECRET=1\n").expect("write blocked env");
    std::fs::write(root.join(".z3-trace"), "").expect("write blocked z3 trace");
    std::fs::create_dir(root.join("node_modules")).expect("create blocked node_modules");
    let matches = blocked_root_files(root, DEFAULT_BLOCKED_ROOT_FILES.iter().copied());
    assert_eq!(
        matches,
        vec![
            ".env".to_string(),
            ".z3-trace".to_string(),
            "node_modules".to_string()
        ]
    );
}

#[test]
fn parse_blocklist_ignores_comments() {
    let tmp = tempdir().expect("tempdir");
    let blocklist = tmp.path().join("blocklist.txt");
    std::fs::write(
        &blocklist,
        "# comment\n\ncustom_placeholder.txt\n  extra.log  \n",
    )
    .expect("write blocklist");
    let blocked = parse_blocklist_file(&blocklist).expect("parse blocklist");
    assert!(blocked.contains("custom_placeholder.txt"));
    assert!(blocked.contains("extra.log"));
    assert_eq!(blocked.len(), 2);
}

#[test]
fn report_passes_when_no_matches() {
    let tmp = tempdir().expect("tempdir");
    let root = tmp.path();
    std::fs::write(root.join("README.md"), "ok\n").expect("write readme");
    std::fs::write(root.join(".env.example"), "SAFE=1\n").expect("write env example");
    let report = build_report(root, DEFAULT_BLOCKED_ROOT_FILES, &Default::default());
    assert!(report.pass);
    assert!(report.matches.is_empty());
    assert_eq!(report.repo_root, Path::new(root).display().to_string());
}
